### ISSUE TITLE
In DNNClassifier evaluation results, the 'loss' is not always = 'average_loss' * batch_size

### ISSUE THREAD

**Issue by ageron** [tensorflow/tensorflow#21557]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOSX 10.13.6
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: N/A
- **TensorFlow installed from (source or binary)**: yes
- **TensorFlow version (use command below)**: v1.10.0-rc1-19-g656e7a2b34 1.10.0
- **Python version**: Python 3.6.6
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**:

### Describe the problem
The evaluation results of a `DNNClassifier` include a `'loss'` and an `'average_loss`'.  It seems that the `'average_loss'` is correct (same as when I compute it manually), while the `'loss'` is weird. 
 It seems that it's roughly equal to the `'average_loss'` times the batch size, but not always. This seems like a bug to me.

I wrote a test (see the code below), and here is its output with different batch sizes (notice that the loss/average_loss ratio is often equal or close to the batch size, but not always, and sometimes it is exactly one off, or simply far from it):

```
Batch size: 1 loss: 0.75871724 average_loss: 0.75871724 Ratio: 1.0
Batch size: 11 loss: 8.33756 average_loss: 0.7587179 Ratio: 10.989012
Batch size: 21 loss: 15.806628 average_loss: 0.75871813 Ratio: 20.833334
Batch size: 31 loss: 22.991457 average_loss: 0.7587181 Ratio: 30.30303
Batch size: 41 loss: 30.348719 average_loss: 0.75871795 Ratio: 40.0
Batch size: 51 loss: 37.935898 average_loss: 0.75871795 Ratio: 50.0
Batch size: 61 loss: 44.63047 average_loss: 0.758718 Ratio: 58.82353
Batch size: 71 loss: 50.581203 average_loss: 0.7587181 Ratio: 66.666664
Batch size: 81 loss: 58.36292 average_loss: 0.75871795 Ratio: 76.92307
Batch size: 91 loss: 68.974365 average_loss: 0.758718 Ratio: 90.90909
Batch size: 101 loss: 75.8718 average_loss: 0.758718 Ratio: 100.0
Batch size: 111 loss: 75.871796 average_loss: 0.75871795 Ratio: 100.0
Batch size: 121 loss: 84.302 average_loss: 0.758718 Ratio: 111.111115
Batch size: 131 loss: 94.83975 average_loss: 0.758718 Ratio: 125.0
Batch size: 141 loss: 94.83975 average_loss: 0.758718 Ratio: 125.0
```

### Source code / logs
```python
from __future__ import division, print_function, unicode_literals

import logging
logging.basicConfig(level=logging.WARNING) # or else it's too verbose

import tensorflow as tf
#tf.enable_eager_execution()  # same exact output if you activate eager execution
import numpy as np

# make the code reproducible (see https://youtu.be/Ys8ofBeR2kA ;-) )
np.random.seed(1234)
config = tf.estimator.RunConfig(tf_random_seed=1234)

num_col = tf.feature_column.numeric_column("X", shape=[10])
dnn_clf = tf.estimator.DNNClassifier(hidden_units=[200, 100], n_classes=3,
                                     feature_columns=[num_col], config=config)

# some random data
X = np.random.randn(1000, 10)
y = np.random.randint(3, size=1000)

# train the model
input_fn = tf.estimator.inputs.numpy_input_fn(
    x={"X": X}, y=y, num_epochs=10, batch_size=32, shuffle=False)
dnn_clf.train(input_fn=input_fn)

# evaluate the model with different batch sizes
for batch_size in range(1, 151, 10):
    input_fn = tf.estimator.inputs.numpy_input_fn(
        x={"X": X}, y=y, batch_size=batch_size, shuffle=False)
    eval_results = dnn_clf.evaluate(input_fn=input_fn)
    print("Batch size:", batch_size, end=" ")
    print("loss:", eval_results["loss"], end=" ")
    print("average_loss:", eval_results["average_loss"], end=" ")
    print("Ratio:", eval_results["loss"] / eval_results["average_loss"])
```

**Comment by ispirmustafa** [comment]
average_loss is loss per example count. and loss is loss per mini batch. As you said, the ratio is roughly `batch_size`. It should be exactly `batch_size` if all `batch_size`s are equal for all mini batches. The given batch sizes in your test, make the last batch not a full one. That's the reason for difference.

**Comment by ageron** [comment]
Oh okay, that's clear, thanks @ispirmustafa , I'll see if I can document this somewhere.  Cheers.
